### PR TITLE
refactor(early-kickout): add `ChunkProducers` DB column with default persistence

### PR DESCRIPTION
**Pull request by stedfn**
- Add new `DBCol::ChunkProducers` column (insert_only, not cold, not GC'd)
- Key: `(prev_block_hash, shard_id)` → `ValidatorStake`
- Wire writes into genesis init, header sync, and block processing
- Add `ProtocolFeature::EarlyKickout` (protocol version 152) — writes gated behind runtime check, no-op on stable
- Always writes default (non-blacklisted) assignments — behavior-preserving
- Add unit tests for genesis, header sync, block processing, sampling parity, and epoch boundary skip

### CODE REVIEW

File: chain/chain/src/garbage_collection.rs [hunk 1]
```diff
@@ -1307,6 +1307,9 @@ impl<'a> ChainStoreUpdate<'a> {
             // because we remove unneeded keys as we add new ones.
             | DBCol::StateSyncHashes
             => unreachable!(),
+            // ChunkProducers is not garbage collected. Once dynamic chunk producer
+            // sampling ships, historical assignments cannot be recomputed.
+            DBCol::ChunkProducers => unreachable!(),
```
Comment (Trisfald) [inline chain/chain/src/garbage_collection.rs:1312]: Is the plan to never garbage collect this column? Or is this a temporary shortcut?

Comment (stedfn) [reply chain/chain/src/garbage_collection.rs:1312]: Just temporary, the feature would not be stabilized without garbage collection implemented. Will update the comment as it is indeed unclear

File: core/store/src/columns.rs [hunk 2]
```diff
@@ -623,7 +630,8 @@ impl DBCol {
             | DBCol::FlatStorageStatus
             | DBCol::EpochSyncProof
             | DBCol::StateSyncHashes
-            | DBCol::StateSyncNewChunks => false,
+            | DBCol::StateSyncNewChunks
+            | DBCol::ChunkProducers => false,
         }
```
Comment (jancionear) [inline core/store/src/columns.rs:627]: Should `is_cold(DBCol::ChunkProducers)` be `false`? Archival nodes might want to find out who was a chunk producer for a given height.
/cc @staffik

File: chain/chain/src/chain.rs [hunk 3]
```diff
@@ -1446,6 +1447,59 @@ impl Chain {
...
+                tracing::debug!(
+                    target: "chain",
+                    height = header.height(),
+                    ?prev_block_hash,
+                    ?e,
+                    "skipping chunk producer save: epoch not available"
+                );
+                return Ok(());
+            }
+        };
```
Comment (jancionear) [inline chain/chain/src/chain.rs]: When does that happen? I thought that we always know what the next epoch is.

File: chain/chain/src/chain.rs [hunk 4]
```diff
@@ -1446,6 +1447,59 @@ impl Chain {
...
+    /// `(header.hash(), shard_id)`. This makes historical chunk producer lookups
+    /// available from the DB without recomputation.
+    ///
+    /// Gated behind `EarlyKickout` protocol feature. No-op when disabled.
+    pub(crate) fn save_chunk_producers_for_header(
+        epoch_manager: &dyn EpochManagerAdapter,
+        header: &BlockHeader,
+        chain_store_update: &mut ChainStoreUpdate,
+        protocol_version: ProtocolVersion,
+    ) -> Result<(), Error> {
```
Comment (jancionear) [inline chain/chain/src/chain.rs]: `Chain`  is an unusual place to put a function like that. Why not put it inside of `ChainUpdate`, like other functions that save data on block postprocessing?

~~The flow where this function modifies `self.chain_update` which later is reused during postprocessing is pretty hard to follow.~~

File: core/store/src/columns.rs (hunk 2 above)
Comment (Trisfald) [reply core/store/src/columns.rs:627]: I guess it's false right now because we never GC'd the hot column

Indeed long term it should be a cold / cloud column

Comment (stedfn) [reply core/store/src/columns.rs:627]: Yes, it will be cold

Comment (jancionear) [reply core/store/src/columns.rs:627]: How about a TODO? It's easy to forget

File: chain/chain/src/chain.rs (hunk 3 above)
Comment (stedfn) [reply chain/chain/src/chain.rs]: good point, I removed it

File: chain/chain/src/store/mod.rs [hunk 5]
```diff
@@ -1864,6 +1866,43 @@ impl<'a> ChainStoreUpdate<'a> {
...
+        for shard_id in shard_layout.shard_ids() {
+            if let Some(validator_id) =
+                epoch_info.sample_chunk_producer(&shard_layout, shard_id, height)
+            {
+                let validator_stake = epoch_info.get_validator(validator_id);
+                let key = get_block_shard_id(prev_block_hash, shard_id);
+                store_update.insert_ser(DBCol::ChunkProducers, &key, &validator_stake);
+            }
+        }
+        self.merge(store_update);
```
Comment (jancionear) [inline chain/chain/src/store/mod.rs]: It looks like the convention is to put the data to be written in a field, and then do the `insert_ser` operations inside of `finalize`. I'm not sure what's the reasoning behind it, but I think it would be good to follow the existing way of doing things. Shreyan might now why it's done this way.

E.g:

```rust
// save the outgoing receipts generated by chunk from block `hash` for shard `shard_id`
pub fn save_outgoing_receipt(
    &mut self,
    hash: &CryptoHash,
    shard_id: ShardId,
    outgoing_receipts: Vec<Receipt>,
) {
    self.chain_store_cache_update
        .outgoing_receipts
        .insert((*hash, shard_id), Arc::new(outgoing_receipts));
}

// finalize
for (chunk_hash, partial_chunk) in &self.chain_store_cache_update.partial_chunks {
    store_update.insert_ser(DBCol::PartialChunks, chunk_hash.as_ref(), partial_chunk);

    for receipts in partial_chunk.prev_outgoing_receipts() {
        for receipt in &receipts.0 {
            save_receipt(&mut store_update, receipt);
        }
    }
}
```

Comment (stedfn) [reply chain/chain/src/store/mod.rs]: Updated, @shreyan-gupta I am curious if there is any specific reason?

Comment (shreyan-gupta) [reply chain/chain/src/store/mod.rs]: Honestly don't know... That's just how it has historically been
